rotary-encoder/rotary_encoder.go

```diff
@@ -0,0 +1,69 @@
+//go:build tinygo
+
+// Implementation based on:
+// https://gist.github.com/aykevl/3fc1683ed77bb0a9c07559dfe857304a
+
+package rotary_encoder
+
+import (
+	"machine"
+	"runtime/volatile"
+)
+
+var (
+	states = []int8{0, -1, 1, 0, 1, 0, 0, -1, -1, 0, 0, 1, 0, 1, -1, 0}
+)
+
+func New(pinA, pinB machine.Pin) *Device {
+	return &Device{pinA: pinA, pinB: pinB, oldAB: 0b00000011}
+}
+
+type Device struct {
+	pinA machine.Pin
+	pinB machine.Pin
+
+	precision int
+
+	oldAB int
+	value volatile.Register32
+	// value int
+}
+
+type Config struct {
+	Precision int
+}
+
+func (enc *Device) Configure(cfg Config) {
+	enc.pinA.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
+	enc.pinA.SetInterrupt(machine.PinRising|machine.PinFalling, enc.interrupt)
+
+	enc.pinB.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
+	enc.pinB.SetInterrupt(machine.PinRising|machine.PinFalling, enc.interrupt)
+
+	if cfg.Precision > 0 {
+		enc.precision = cfg.Precision
+	} else {
+		enc.precision = 4
+	}
+}
+
+func (enc *Device) interrupt(pin machine.Pin) {
+	aHigh, bHigh := enc.pinA.Get(), enc.pinB.Get()
+	enc.oldAB <<= 2
+	if aHigh {
+		enc.oldAB |= 1 << 1
+	}
+	if bHigh {
+		enc.oldAB |= 1
+	}
+	enc.value.Reg += uint32(states[enc.oldAB&0x0f])
+	// enc.value += int(states[enc.oldAB&0x0f])
+}
+
+func (enc *Device) Value() int {
+	return int(enc.value.Reg) / enc.precision
```

Comment:
This doesn't help. You need to call the `.Get()` and `.Set(value)` methods to access the value in a volatile way.

I realize there isn't a lot of documentation on that package, but this might help: https://tinygo.org/docs/concepts/compiler-internals/volatile/

Reply:
Ok cool thanks, I was going to ask about that because the program still wasn't working without the `time.Sleep` in there

Reply:
Switching to use the Get/Set methods worked, the program functions as expected without the `time.Sleep`